flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/AbstractAlternatingAlignedBarrierHandlerAction.java

```diff
@@ -0,0 +1,92 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.io.checkpointing;
+
+import org.apache.flink.runtime.checkpoint.CheckpointException;
+import org.apache.flink.runtime.checkpoint.channel.InputChannelInfo;
+import org.apache.flink.runtime.io.network.api.CheckpointBarrier;
+import org.apache.flink.runtime.io.network.partition.consumer.CheckpointableInput;
+
+import java.io.IOException;
+
+/**
+ * Actions to be taken when processing aligned checkpoints and possibly switching to unaligned
+ * checkpoints.
+ */
+abstract class AbstractAlternatingAlignedBarrierHandlerAction implements BarrierHandlerAction {
+
+    private final AlignedCheckpointState state;
+
+    protected AbstractAlternatingAlignedBarrierHandlerAction(AlignedCheckpointState state) {
+        this.state = state;
+    }
+
+    @Override
+    public final BarrierHandlerAction alignmentTimeout(
+            Context context, CheckpointBarrier checkpointBarrier)
+            throws IOException, CheckpointException {
+        state.prioritizeAllAnnouncements();
+        CheckpointBarrier unalignedBarrier = checkpointBarrier.asUnaligned();
+        context.triggerTaskCheckpoint(unalignedBarrier);
+        for (CheckpointableInput input : state.getInputs()) {
+            input.checkpointStarted(checkpointBarrier);
+        }
+        context.triggerGlobalCheckpoint(unalignedBarrier);
+        return new CollectingBarriersUnaligned(true, state.getInputs());
+    }
+
+    @Override
+    public final BarrierHandlerAction announcementReceived(
+            Context context, InputChannelInfo channelInfo, int sequenceNumber) {
+        state.addSeenAnnouncement(channelInfo, sequenceNumber);
+        return this;
+    }
+
+    @Override
+    public final BarrierHandlerAction barrierReceived(
+            Context context, InputChannelInfo channelInfo, CheckpointBarrier checkpointBarrier)
+            throws IOException, CheckpointException {
+        if (checkpointBarrier.getCheckpointOptions().isUnalignedCheckpoint()) {
+            BarrierHandlerAction unalignedState = alignmentTimeout(context, checkpointBarrier);
+            return unalignedState.barrierReceived(context, channelInfo, checkpointBarrier);
+        }
+
+        state.removeSeenAnnouncement(channelInfo);
+        state.blockChannel(channelInfo);
+        if (context.allBarriersReceived()) {
+            context.triggerGlobalCheckpoint(checkpointBarrier);
+            state.unblockAllChannels();
+            return new AlternatingWaitingForFirstBarrier(state.getInputs());
+        } else if (context.isTimedOut(checkpointBarrier)
+                || checkpointBarrier.getCheckpointOptions().isUnalignedCheckpoint()) {
+            return alignmentTimeout(context, checkpointBarrier);
+        }
+
+        return transitionAfterBarrierReceived(state);
+    }
+
+    protected abstract BarrierHandlerAction transitionAfterBarrierReceived(
```

Comment:
It feels a bit odd that you pass state as a parameter here. I guess it's needed to convert to a different action? How about adding some protected methods here like `toAlternating`, `toAligned`, etc. and instead of passing the state just call these methods in `transitionAfterBarrierReceived` implementations.

Btw I'd call this method `convertAfter...`. Transition makes sense in a state machine, which is hardly recognizable here.

Reply:
It's not about switching to alternating or aligned. It's about if we originate from `WaitingForFirstBarrier` or `CollectingBarriers` states. The only difference in `barrierReceived` method for those two states is which state we convert to.